debian/bsesh.1
.TH BSESH 1 "29 Mar 2003" "Debian GNU/Linux" "BEAST/BSE"
.SH NAME
bsesh \- a scheme shell for BEAST/BSE
.SH SYNOPSIS
.B bsesh [OPTIONS] ...
.SH DESCRIPTION
.B bsesh
is a scheme interpreter based on guile that interacts with
.B beast.
It is usually launched by
.B beast
itself.

.SH OPTIONS
.IP --bse-command-pipe <input> <output> <expression>
Provide the input and output pipes from
.B beast
and a Scheme expression to evaluate.
.IP --bse-enable-server
Activate the
.B beast
server.
.IP --bse-enable-register
Register
.B beast
scripts.
.IP -p
Register
.B beast
plugins.
.IP --g-fatal-warnings
Cause runtime warnings to immediately abort the program.
.IP -l FILE
Load scheme source code from file.
.IP -e FUNCTION
After reading script, apply FUNCTION to command-line arguments.
.IP -ds
do -s SCRIPT at this point (note that this argument must be used in
conjuction with -s).
.IP --help
Describe command line options and exit.
.IP --debug
Start bsesh with debugging evaluator and backtraces enabled
(useful for debugging guile scripts).
.IP --version
Display guile version and exit.
.IP --emacs
Enable emacs protocol for use from within emacs (experimental).
.IP --
Stop argument processing, start bsesh in interactive mode.
.IP -c EXPR
Stop argument processing, evaluate EXPR as a scheme expression.
.IP -s SCRIPT-FILE
Load Scheme source from SCRIPT-FILE and execute as a script.

.SH AUTHOR
This manpage has been written by Sam Hocevar <sam@zoy.org> for the Debian
GNU/Linux system, based on the guile manpage written by Robert Merkel
<rgmerk@mira.net>.
.SH "SEE ALSO"
beast(1), guile(1).



NEWS
Overview of Changes in BEAST/BSE 0.9.0:		(2014-10-28)

* Fixed missing includes. [stw]
* Updates to build with Rapicorn1410.
* Work around GClosure API breakages in GLib.
* Work around GLib API deprecations.
* Remove old autotools and libtool workarounds.
* Updates to compile with g++-4.9.1.
* Fix issues on 32bit platforms.
* Optimize ELF relocations.
* Dependency update to Python 2.7.5.
* Updates to the manual pages.
* Improved utilization of compiler vectorizations.

Overview of Changes in BEAST/BSE 0.8.2:		(2013-07-06)

* Introduced Rapicorn1307 dependency.
* Establish Aida remote interfacing between Beast and BSE.
* Removed birnet/ sublibrary entirely, replaced with Rapicorn uses.
* Added support for loading and saving of FLAC encoded samples. [stw]
* Ported code to use Rapicorn's type safe cxxprintf instead of vprintf.
* Lots of documentation improvements and fixes.
* Automatization of documentation builds documentation uploads.
* Beast now ships complete documentation builds in tarballs.


Overview of Changes in BEAST/BSE 0.8.0:		(2013-03-25)

* Added use of the Rapicorn library, require version 13.0.0.
* Adapted IDL types to other projects.
* Adapted code to make use of C++11 and STL threads.
* Various performance improvements through EventFD, Spinlocks, lambdas, etc.
* Major documentation improvements, see: http://dev.testbit.eu/beast/latest/
* Fixed printf format-security issues (#665097). [Alessio Treglia]
* Lots of code cleanups, licensing cleanups.


Overview of Changes in BEAST/BSE 0.7.8:

* Wave oscillator fixes for 96k samples (bug #612281). [stw]
* Performance improvements of the scheme shell. [stw]
* Allow Alt+MouseButton1 to drag synthesis modules.
* Merged plugin libraries, improved build times.
* Use the ALSA 'default' PCM driver by default.
* Adapt code to allow fftw drop ins. [stw]
* Migrated code base to compile as C++11.
* Lots of simplifications and cleanups.
* Extended MIDI device listing.
* Improved Beast startup time.


Overview of Changes in BEAST/BSE 0.7.6:

* Packaging fixups, download and integrate documentation from beast wiki.
* Work around regressions introduced in recent GLib and GTK+ versions.
* Various bug and leak fixups.
* Moved lots of code to C++.
* Build documentation with doxygen.
* Code fixups for g++-4.4, g++-4.7.
* Improved DavOrgan frequency handling.


Overview of Changes in BEAST/BSE 0.7.4:

* Renamed the project to Better Audio System / Better Sound Engine
* Moved project website to: http://beast.testbit.eu/
* Various build system fixes [stw,timj]
* License fixups for some scripts [stw]
* Fixed subnormal tests on AMD64 if SSE unit is in DAZ mode [stw]
* Replaced slow resampler checks with a much faster resampling test [stw]
* Performance improvements for various tests [stw]
* GLib 2.28 unit test porting [stw]
* Speed improvements for record field name [stw]
* Fixed XRUNs in ALSA driver on 64bit systems [timj]
* Added beast.doap [Jonh Wendell]
* PO handling improvements.
* Updated German translation.
* Updated Norwegian bokmål translation [Kjartan Maraas]
* Added e-Telugu translation [Veeven]


Overview of Changes in BEAST/BSE 0.7.2:

* Moved Beast/BSE to GNU LGPL, use AS-IS license for examples
* Module changes and additions:
  ArtsCompressor       - Relicensed to LGPL with permission from Matthias Kretz
  BseContribSampleAndHold - Relicensed to LGPL with permission from Artem Popov
  DavXTalStrings          - Use deterministic random numbers for unit tests
  BseNoise                - Improved random number generator
* Switched to autogenerated ChangeLogs
* Error bell can be muted in beast preferences dialog
* Added multisample creation/editing command line tool: bsewavetool
* Support adjustable volume, pitching and drum envelopes in .bsewave files
* Added Retro Acoustic drum kit [Tim, Stefan]
* New loadable Instruments/Effects:
  - BQS Bass Drum E8012 [Tim, Stefan]
  - BQS Slow Hum [Stefan, William DeVore]
  - FSM Fresh Water Bass instrument [Krzysztof Foltman]
  - FSM Growl Bass instrument [Krzysztof Foltman]
  - FSM Synth String Sweep [Krzysztof Foltman]
* Added support for loading 32bit and 24bit PCM-format WAV files
* Added support for gcc-4.4 and automake-1.10
* Added support for guile-1.8, guile-1.6 remains as minimum requirement
* Various fixes, improvements and much improved test coverage.
* Bug fixes: #452604, #468229, #344388, #451086, #450724, #454121, #491552,
             #450490, #441936, #336766, #433431, #474332, #474244, #456879,
             #456408, #424897 [Tim Janik, Stefan Westerfeld]
* Migrated translation support to use awk, sed and po/Makefile.am.
* Updated German translation [Mario Blättermann]
* Updated Italien translation [Michele Petrecca]
* Updated Occitan translation [Yannig Marchegay]
* Updated Brazilian Portugues translation [Leonardo Ferreira Fontenelle]
* Updated British English translation [David Lodge]
* Updated Spanish translation [Jorge Gonzalez]
* Updated Slovenian translation [Andrej Znidarsic]
* Updated Danish translation [Joe Hansen]
* Updated French translation [Bruno Brouard]
* Added Norwegian bokmal translation [Kjartan Maraas]
* Added Ukrainian translation [Maxim V. Dziumanenko]


Overview of Changes in BEAST/BSE 0.7.1:

* Fixed SUID security vulnerability by validating success of seteuid/setreuid,
  related security advisories, describing the vulnerability:
    http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2006-2916 # artswrapper
    http://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2006-4447 # X.Org
* Added factor 2 SSE-optimized up/down resampler [Stefan Westerfeld]
* Added support for varying musical tuning systems:
  - 12-TET, 7-TET, 5-TET,
  - Diatonic-Scale, Indian-Scale, Pythagorean-Tuning,
  - Pentatonic-5-Limit, Pentatonic-Blues, Pentatonic-Gogo,
  - Quarter-Comma-Meantone, Silbermann-Sorge,
  - Werckmeister-3, Werckmeister-4, Werckmeister-5, Werckmeister-6,
    Kirnberger-3 and Young
  [Hanno Behrens, Tim Janik]
* Module changes and additions:

shell/bsesh.c
/* BSE-SCM - Bedevilled Sound Engine Scheme Wrapper
 * Copyright (C) 2002 Tim Janik
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
 */
#define G_LOG_DOMAIN "BseShell"

#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <guile/gh.h>
#include <bse/bse.h>
#include "bsescminterp.h"
#include "topconfig.h"

#define	PRG_NAME	"bsesh"

#define BSE_EXIT_STATUS 102

#define	BOILERPLATE_SCM		BSE_PATH_SCRIPTS "/bse-scm-glue.boot"


/* --- prototypes --- */
static void	gh_main			(gint	 argc,
					 gchar	*argv[]);
static void	shell_parse_args	(gint    *argc_p,
					 gchar ***argv_p);


/* --- variables --- */
static gint            bse_scm_pipe[2] = { -1, -1 };
static gchar          *bse_scm_eval_expr = NULL;
static gboolean        bse_scm_enable_register = FALSE;
static gboolean        bse_scm_register_plugins = TRUE;
static SfiComPort     *bse_scm_port = NULL;
static SfiGlueContext *bse_scm_context = NULL;


/* --- functions --- */
static void
port_closed (SfiComPort *port,
	     gpointer    close_data)
{
  /* we don't do anything fancy here */
  if (port)
    exit (BSE_EXIT_STATUS);
}

int
main (int   argc,
      char *argv[])
{
  const gchar *env_str;

  g_thread_init (NULL);
  g_set_prgname (PRG_NAME);
  sfi_init ();

  env_str = g_getenv ("BSE_SHELL_SLEEP4GDB");
  if (env_str && atoi (env_str) > 0)
    {
      g_message ("going into sleep mode due to debugging request (pid=%u)", getpid ());
      g_usleep (2147483647);
    }

  shell_parse_args (&argc, &argv);

  if (bse_scm_pipe[0] >= 0 && bse_scm_pipe[1] >= 0)
    {
      bse_scm_port = sfi_com_port_from_pipe (PRG_NAME, bse_scm_pipe[0], bse_scm_pipe[1]);
      sfi_com_port_set_close_func (bse_scm_port, port_closed, NULL);
      if (!bse_scm_port->connected)
	{
	  g_printerr ("%s: failed to connect to pipe (%d, %d)\n", PRG_NAME, bse_scm_pipe[0], bse_scm_pipe[1]);
	  exit (BSE_EXIT_STATUS);
	}
      bse_scm_context = sfi_glue_encoder_context (bse_scm_port);
    }

  if (!bse_scm_context)
    {
      /* start our own core thread */
      bse_init_async (&argc, &argv, NULL);
      bse_scm_context = bse_init_glue_context (PRG_NAME);
    }

  gh_enter (argc, argv, gh_main);

  return 0;
}


static void
gh_main (int   argc,
	 char *argv[])
{
  /* initial interpreter setup */
  if (bse_scm_enable_register)
    bse_scm_enable_script_register (TRUE);
  else
    bse_scm_enable_server (TRUE);
  sfi_glue_context_push (bse_scm_context);

  /* initialize interpreter */
  bse_scm_interp_init ();

  /* exec Bse Scheme bootup code */
  gh_load (BOILERPLATE_SCM);

  /* eval or interactive */
  if (bse_scm_eval_expr)
    gh_eval_str (bse_scm_eval_expr);
  else
    gh_repl (argc, argv);

  /* shutdown */
  sfi_glue_context_pop ();
  if (bse_scm_port)
    {
      sfi_com_port_set_close_func (bse_scm_port, NULL, NULL);
      sfi_com_port_close_remote (bse_scm_port, FALSE);
    }
  sfi_glue_context_destroy (bse_scm_context);
}

static void
shell_parse_args (gint    *argc_p,
		  gchar ***argv_p)
{
  guint argc = *argc_p;
  gchar **argv = *argv_p;
  guint i, e;

  for (i = 1; i < argc; i++)
    {
      if (strcmp (argv[i], "--") == 0)
	break;
      else if (strcmp (argv[i], "--g-fatal-warnings") == 0)
	{
	  GLogLevelFlags fatal_mask;

	  fatal_mask = g_log_set_always_fatal (G_LOG_FATAL_MASK);
	  fatal_mask |= G_LOG_LEVEL_WARNING | G_LOG_LEVEL_CRITICAL;
	  g_log_set_always_fatal (fatal_mask);
	  argv[i] = NULL;
	}
      else if (strcmp (argv[i], "--bse-pipe") == 0)
	{
	  /* first two arguments are input and output pipes from BseServer */
	  bse_scm_pipe[0] = -1;
	  bse_scm_pipe[1] = -1;
	  if (i + 2 < argc)
	    {
	      bse_scm_pipe[0] = atoi (argv[i + 1]);
	      bse_scm_pipe[1] = atoi (argv[i + 2]);
	      argv[i++] = NULL;
	      argv[i++] = NULL;
	    }
	  argv[i] = NULL;
	  if (bse_scm_pipe[0] < 2 || bse_scm_pipe[1] < 2)
	    {
	      g_printerr ("%s: invalid arguments supplied for: --bse-pipe <inpipe> <outpipe>\n", PRG_NAME);
	      exit (BSE_EXIT_STATUS);
	    }
	}
      else if (strcmp (argv[i], "--bse-eval") == 0)
	{
	  bse_scm_eval_expr = NULL;
	  if (i + 1 < argc)
	    {
	      bse_scm_eval_expr = argv[i + 1];
	      argv[i++] = NULL;
	    }
	  argv[i] = NULL;
	  if (!bse_scm_eval_expr)
	    {
	      g_printerr ("%s: invalid arguments supplied for: --bse-eval <expression>\n", PRG_NAME);
	      exit (BSE_EXIT_STATUS);
	    }
	}
      else if (strcmp (argv[i], "--bse-enable-register") == 0)
	{
	  bse_scm_enable_register = TRUE;
	  argv[i] = NULL;
	}
      else if (strcmp (argv[i], "--bse-no-plugins") == 0)
	{
	  bse_scm_register_plugins = FALSE;
	  argv[i] = NULL;
	}
    }

  e = 0;
  for (i = 1; i < argc; i++)
    {
      if (e)
	{
	  if (argv[i])
	    {
	      argv[e++] = argv[i];
	      argv[i] = NULL;
	    }
	}
      else if (!argv[i])
	e = i;
    }
  if (e)
    *argc_p = e;
}
